src/Search/Queue/WatchableQueue.php

```diff
@@ -0,0 +1,46 @@
+<?php
+
+namespace eLife\Search\Queue;
+
+interface WatchableQueue
+{
+    /**
+     * Adds item to the queue.
+     *
+     * Mock: Add item to queue.
+     * SQS: This will set the queue item into the memory slot for re-processing.
+     */
+    public function enqueue(QueueItem $item) : bool;
+
+    /**
+     * Starts process of removing item.
+     *
+     * Mock: Move to separate "in progress" queue.
+     * SQS: this will change the timeout of the in-memory item.
+     */
+    public function dequeue(int $timeout = 30) : QueueItem;
+
+    /**
+     * Commits to removing item from queue, marks item as done and processed.
+     *
+     * Mock: Remove item completely.
+     * SQS: this will delete the item from the queue.
+     */
```

Comment:
There is also the option of committing the message as deleted from the queue after the job is added to Gearman, rather than waiting for the whole job to be executed. That would free you from having to notify SQS (not sure how too) when the Gearman job is finished; it would only work reliably if we persist Gearman jobs into a database (IIRC MySQL can be configured) rather than keeping them in memory, to avoid data loss in case of crash.

About the process being terminated, the default behavior of a PHP process is to stop immediately; what happens is there will be no `commit()` call happening for the dequeued item, and after its visibility timeout expires it will be accessible to the next process taking items off the queue. This is at-least-once delivery and it's safe. It can be optimized by intercepting the SIGTERM signal and setting the visibility timeout to 0 when you receive it, but PHP works in a single thread so that would mean putting the `pcntl_signal_dispatch()` call that triggers (intercepted) signals delivery somewhere in the loop... it gets complex quickly.